### PR TITLE
Tagged unions

### PR DESCRIPTION
**Pull request by xogeny**
Here is an example using tagged unions.  I've gotten rid of references to `instanceof`
(for reasons I mentioned before).  Note that the definitions in `AST.ts` might seem
a little scary and verbose at first.  But I decided to err on the side of being overly 
explicit.  At a minimum, I think it leads to better documentation.

I've tried to follow some best practices that lead to the greatest degree of type
checking.  For example, you'll find that I refined the original atomic types into
individual versions for each type of node (instead of relying on `value: any`).  Again,
I just thought it would be better to be explicit.  There is an added benefit that more
type checking will be applied.

Let me know what you think.

### CODE REVIEW

**Comment by sgarciac** [comment]
Thanks for the PR, I'll have a closer look this week

**Comment by xogeny** [comment]
FYI, when I merge my Jest work into this branch and run, I see exactly the same four errors.  So it seems as though my "tagged union" changes don't break anything.  The ideal case would be to create a set of tests that pass 100% on your `master` version first and then show that the same tests pass 100% on my tagged union version.  I'll see if I can find an error in your code that converts between representations.

**Comment by xogeny** [comment]
My assessment is that the `toml-test` baselines are actually broken for table arrays.  I say this because the TOML spec implies that this:

```toml
points = [ { x = 1, y = 2, z = 3 },
           { x = 7, y = 8, z = 9 },
           { x = 2, y = 4, z = 8 } ]
```

Should be equivalent to:

```toml
[[points]]
x = 1
y = 2
z = 3

[[points]]
x = 7
y = 8
z = 9

[[points]]
x = 2
y = 4
z = 8
```

But `toml-test` seems inconsistent in its deserialized JSON representation.  It seems to represent the former as:

```json
{
  "points": {
    "type": "array",
    "value": [{ "x": 1, "y": 2, "z": 3 }, { "x": 7, "y": 8, "z": 9 }, { "x": 2, "y": 4, "z": 8 } ]
  }
}
```

(see `array.json` for why I think that...since none of the cases in `toml-test` actually test arrays of inline tables)

On the other hand, it appears it would translate the latter as:

```json
{
  "points": [{ "x": 1, "y": 2, "z": 3 }, { "x": 7, "y": 8, "z": 9 }, { "x": 2, "y": 4, "z": 8 } ]
}
```

(see `table-array-one.json` for why I would think that).

**Comment by xogeny** [comment]
Now with minty fresh tests (see #2).  All tests vs. TOML spec and `toml-test` pass.

File: src/AST.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import ct = require('chevrotain');
+
+// Table headers. We keep the Token for reporting errors during the loading phase.
+export const tableHeader: 'tableHeader' = 'tableHeader';
+export class TomlTableHeader {
```
Comment (sgarciac) [inline src/AST.ts]: should be an interface?

File: src/main.ts [hunk 2]
```diff
@@ -40,8 +41,8 @@ function bombadilToTomlTestAtomicValue(input: toml.TomlAtomicValue) {
     }
 }
```
Comment (sgarciac) [inline src/main.ts]: we should either copy your new fixed version of bombadilToTomlTest here, of get rid off this file and the scripts. I dont think toml-test is being mantained, so prefer the later option.

**Comment by sgarciac** [comment]
Everything seems quite nice. I sent you an invitation as a collaborator to the repository. If you are ok with it, feel free to merge your changes. Otherwises, juste me le know and I'll merge them myself.

**Comment by xogeny** [comment]
I can do the merge.  Only wrinkle...I actually do my dev work as @mtiller (long story).  So can you add @mtiller as a collaborator?  Then I should be able to merge.  I've got all the changes ready to go.  BTW, I added my repo forked repo to CircleCI and it works.  So you can trivially add CircleCI testing to the repo.

**Comment by xogeny** [comment]
I'm closing this PR because #5 is a much better PR (rolls everything up together).

One odd thing...I notice that, as @xogeny, I cannot merge a PR.  Not sure what is going on there.  I may have other contributions in the future (once I get back to my TOML idea), so keeping @xogeny (for web-based operations like merging PRs) and @mtiller (for pushing via command line) as collaborators would be useful.

**Pull request closed without merging**